src/pseudo/PspUpf.jl

```diff
@@ -0,0 +1,200 @@
+using LinearAlgebra
+using Interpolations: linear_interpolation
+using PseudoPotentialIO: load_upf
+
+struct PspUpf{T,IT} <: NormConservingPsp
+    Zion::Int                             # Ionic charge (Z - valence electrons)
+    lmax::Int                             # Maximal angular momentum in the non-local part
+
+    rgrid::Vector{T}                      # Radial grid
```

Comment:
In the UPF file, there's not really a consistent precision for the floats; QE reads them in as doubles (`Float64`), so I guess we could hard-code them that way.

Reply:
They're very likely produced as Float64, so makes sense to read them as such. Then you just have to make sure that they are converted to the DFTK working precision when the eval_ functions are called

Reply:
I disagree here. We might want to do AD through at least some of these parameters, so having the type hard-coded is not a good idea.

Reply:
We don't AD through the constructor: this is just a memory representation of the file, and it's confusing to pretend otherwise

Reply:
> We don't AD through the constructor: 

Oh yes we do. That's the easiest way to do AD here. We construct a PSP object using our custom data and build the model from it.

Reply:
Wouldn't this be handled through a type produced by the (Julia) pseudo generator? Here, we're simply adding support for the UPF file format.

Reply:
Ok, depends on your point of view. For me `PspHgh` does not reflect a file format, but rather a pseudopotential format. With `PspUpf` as we implement it now this sort of broke, because it more is the front-end for a file format, I agree. You can, however, also view this as yet another psp model, where you have a bunch of parameters to specify.

I'm not sure what's the best point of view to have here, we should discuss this in the light of the more general PSP interface. For the moment, however, I'd argue from a user point of view the easiest to understand the PSP influence on the computational result, one would vary the parameters in the PspUpf struct a little and see what happens. At that point you need them to be able to support non-float types.

Reply:
For me PspUpf definitely reflects the file format (or, if you want, the general concept of a pseudopotential given on a grid; we could rename it to PspNumerical if you want?). You probably don't want to modify the grid values directly, at least until we do reverse diff... The Psp interface is lightweight enough that if you want to do AD with a pseudopotential in a simple functional form, you can just roll your own

Reply:
It would be very nice to have a clean and simple `PspNumerical`, but the devil is in the details. For example, spin-orbit coupling is handled so differently between UPF and PSP8 that Abinit doesn't support it (likely because Abinit implements spin-orbit quite differently than QE).

Because of that, I think it will prove to be much easier to handle these different numeric psp formats with a simpler file-to-struct + interface approach (at least for now). We might run into issues with this approach when/if we look to support e.g. ultrasoft pseudos where the file type is the same but the pseudo contained is conceptually different.

I also agree with Antoine that it doesn't make a lot of sense at the moment to take derivatives w.r.t. quantities on the grid with forward mode AD. I would even argue that taking derivatives w.r.t. quantities on the grid is generally not a good idea. It would be much better to take derivatives w.r.t. pseudo generation parameters through a generator written in Julia (which I guess is in @unkcpz 's plans).

Reply:
Also, does it make sense to do AD w.r.t. the values on the grid when they're constructed conserve the norm of the all-electron wavefunction? 

An example: the best fix to problems we were seeing in Barium systems for the common workflows project was to add an f-projector to the pseudo. I'm not sure this can be done with AD on numeric pseudos unless you manually add a guess f-channel (and meet the norm-conserving constraint of the all-electron wavefunction -- which in most cases we don't have access to from the file).

Reply:
I agree on the grid aspect and the idea with the derivatives wrt. generator. But keep in mind that when you want to do that using forward-diff (for e.g. a few hyperparameters that influence your optimisation) you still need to be able to store duals on the grid, so a hard-coded Float will not be sufficient. I'm ok to specialise on Float for now if you think it simplifies the code. Always better to generalise when we know what we actually want.

But I don't resonate well with the idea to have a special struct just for taking derivatives, because from the user perspective I select my PSP and then I'm curious to understand what the influence of that choice on my computational result is. So I ask for a gradient ... and that should not involve manual fiddeling.